### PR TITLE
[PFX-543] - add @gasket/plugin-typescript, adjust types to pass typecheck

### PR DESCRIPTION
**Pull request by mmason2-godaddy**
## Note

I'm aware the generated `tsconfig.json` may need to change upon further testing. This is reflected in a follow up ticket. We need to merge a lot of pending work and publish a beta release so we can reliably test these changes.

## TS Support with Gasket Actions

Given the variety of Gasket app recipes and the different ways Typescript can be used, we need to identify a way to support Typescript in Gasket apps that is flexible and can be used in a variety of scenarios.

### Step One: Preset Prompts

Each preset needs to own the prompt for Typescript support. When Typescript is selected, the preset will add that setting to the `context` object.

### Step Two: CreateContext

The `createContext` will contain a property for Typescript support, potentially named `typescript` which could be a simple boolean value. A plugin needs to hook the `create` lifecycle, read the `createContext` and add Typescript support to the app. It's not obvious which plugin should be responsible for this. Given that Typescript support is not related to most of our current plugins adding Typescript specific logic would pollute the plugins with unrelated code. It would be better to create a new plugin that is responsible for adding Typescript support. This plugin could run last in the `create` lifecycle.

### Step Three: @gasket/plugin-typescript

This plugin will be create only and will be responsible for adding scripts, files and dependencies to the app. It will also add a `tsconfig.json` file to the app.

```js
// Files
// generator/tsconfig.json
files.add(`generator/*`);

// Scripts
pkg.scripts({
  "build": "tsc",
  "start": "tsc && node dist/server.js",
  "dev": "nodemon server.ts"
});

pkg.add('devDependencies', {
  "typescript": "^3.7.2",
  "ts-node": "^8.6.2",
  "nodemon": "^2.0.2"
});

```

### Step Four: Bring it all together

If a preset allows for the Typescript option and the prompt is selected, the `createContext` will contain the `typescript` property. In the `presetConfig` lifecycle of the given preset, the `@gasket/plugin-typescript` plugin will be added to the app and the create hook will be executed.

```js
import pluginExpress from '@gasket/plugin-express';
import pluginTypescript from '@gasket/plugin-typescript';
import pluginHttps from '@gasket/plugin-https';
import pluginWinston from '@gasket/plugin-winston';
import pluginDocs from '@gasket/plugin-docs';
import pluginDocusaurus from '@gasket/plugin-docusaurus';
import pluginResponeData from '@gasket/plugin-response-data';

// Example of API preset where explicitly adding the @gasket/plugin-typescript is necessary
export default {
  name: 'gasket-api-preset',
  hooks: {
    async presetPrompts(gasket, createContext, { prompt }) {
      const answers = await prompt([
        {
          name: 'typescript',
          message: 'Do you want to use TypeScript?',
          type: 'confirm'
        }
      ]);

      return { ...createContext, ...answers };
    },
    async presetConfig(gasket, createContext) {
      return {
        plugins: [
          pluginExpress,
          pluginHttps,
          pluginWinston,
          pluginDocs,
          pluginDocusaurus,
          pluginResponeData,
          createContext.typescript ? pluginTypescript : null
        ].filter(Boolean)
      }
    }
  }
}

```

### About Next.js

The Next.js Typescript support is a bit different. The `@gasket/plugin-typescript` package will not be include. There is configuration changes that can be made to update the `next.config.js` file to support Typescript. This will be a separate ticket.

### Additional tickets

- [x] [Remove plugin start](https://godaddy-corp.atlassian.net/browse/PFX-567)
- [x] [Port create-gasket-app to ESM](https://godaddy-corp.atlassian.net/browse/PFX-568)
- [x] [Port CLI to ESM](https://godaddy-corp.atlassian.net/browse/PFX-569)
- [x] [Port Engine to ESM](https://godaddy-corp.atlassian.net/browse/PFX-570)
- [x] [TS support for Next.js](https://godaddy-corp.atlassian.net/browse/PFX-571)
- [x] [Determine TS types cli vs create-gasket-app](https://godaddy-corp.atlassian.net/browse/PFX-572)
- [x] [Update tsconfig in the TS plugin if necessary](https://godaddy-corp.atlassian.net/browse/PFX-574)

## Changelog
- Initial release of `@gasket/plugin-typescript`

## Test Plan
Tests passing.

### CODE REVIEW

File: packages/create-gasket-app/lib/index.d.ts [hunk 1]
```diff
@@ -1,261 +1,7 @@
...
-
-  /** gasket.config builder */
-  gasketConfig: ConfigBuilder<GasketConfigFile>;
-
-  // Added by `create-hooks`
-
-  /** Use to add files and templates to generate */
-  files: Files;
-}
+import type { CreateContext } from '@gasket/cli';
```
Comment (agerard-godaddy) [inline packages/create-gasket-app/lib/index.d.ts:4]: I think the CreateContext type should live in create-gasket-app now, right?

File: packages/gasket-cli/lib/index.d.ts [hunk 2]
```diff
@@ -2,6 +2,11 @@ import type { GasketConfigDefinition, MaybeAsync } from '@gasket/engine';
 import type { PackageManager } from '@gasket/utils';
 import type { Command } from 'commander';
 
+/**
+ * https://godaddy-corp.atlassian.net/browse/PFX-572
```
Comment (agerard-godaddy) [inline packages/gasket-cli/lib/index.d.ts:6]: At least for create, the CLI does't play a part in anymore. So if you're going to make related type changes in this PR, move ownership of the types to `create-gasket-app`

File: packages/gasket-plugin-nextjs/lib/create.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+   */
+  handler: function create(gasket, context) {
+    const { files, pkg, testPlugin, addSitemap, nextServerType } = context;
+    const generatorDir = `${__dirname}/../generator`;
+
+    createAppFiles({ files, generatorDir });
+    createTestFiles({ files, generatorDir, testPlugin });
+    createNextFiles({ files, generatorDir });
+    addDependencies({ pkg });
+    addNpmScripts({ pkg, nextServerType: nextServerType });
```
Comment (agerard-godaddy) [inline packages/gasket-plugin-nextjs/lib/create.js:141]: 👍 much easier to digest

File: packages/gasket-plugin-nextjs/lib/create.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+      'start': 'next start',
+      'start:local': 'next start & GASKET_ENV=local node server.mjs',
+      'local': 'next dev & nodemon server.mjs'
+    },
+    'next-custom': {
+      'build': 'next build',
+      'start': 'next build && node server.mjs',
+      'start:local': 'next build && GASKET_ENV=local node server.mjs',
+      'local': 'GASKET_DEV=1 GASKET_ENV=local nodemon server.mjs'
+    }
```
Comment (agerard-godaddy) [inline packages/gasket-plugin-nextjs/lib/create.js:112]: With `createContext.nextServerType`, make your prompt values and keys here clearer

```suggestion
    defaultServer: {
      'build': 'next build',
      'start': 'next start',
      'start:local': 'next start & GASKET_ENV=local node server.mjs',
      'local': 'next dev & nodemon server.mjs'
    },
    customServer: {
      'build': 'next build',
      'start': 'next build && node server.mjs',
      'start:local': 'next build && GASKET_ENV=local node server.mjs',
      'local': 'GASKET_DEV=1 GASKET_ENV=local nodemon server.mjs'
    }
```

File: packages/gasket-plugin-nextjs/lib/prompt.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  if (!('nextServerType' in context)) {
+    const { nextServerType } = await prompt([
+      {
+        name: 'nextServerType',
+        message: 'Which server type would you like to use?',
+        type: 'list',
+        choices: [
+          { name: 'Next Server(CLI)', value: 'next-cli' },
+          { name: 'Custom Next Server', value: 'next-custom' }
+        ]
```
Comment (agerard-godaddy) [inline packages/gasket-plugin-nextjs/lib/prompt.js:13]:
```suggestion
        name: 'nextServerType',
        message: 'Which server type do you want for Next.js?',
        type: 'list',
        choices: [
          { name: 'Next.js Built-in Server', value: 'defaultServer' },
          { name: 'Custom Server', value: 'customServer' }
        ],
        default: 'defaultServer'
```

File: packages/gasket-plugin-nextjs/lib/prompt.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+    newContext.nextServerType = nextServerType;
+  }
+
+  if (!('addSitemap' in context)) {
+    const { addSitemap } = await prompt([
+      {
+        name: 'addSitemap',
+        message: 'Do you want to add a sitemap?',
+        type: 'confirm'
```
Comment (agerard-godaddy) [inline packages/gasket-plugin-nextjs/lib/prompt.js:25]: We should be defaulting this to false:
```suggestion
        type: 'confirm',
        default: false
```

**Review by kbader-godaddy — APPROVED** [review]

**Review by agerard-godaddy — APPROVED** [review]

**Pull request merged**